### PR TITLE
Support partial row in read/write process for partial update.

### LINKED ISSUE #2059: Support partial row in read/write process for partial update
## Feature request

**Is your feature request related to a problem? Please describe.**
Since partial update syntax supported in #1546 , #1829 , need to implement it in read/write process.

**Describe the solution you'd like**
Use partial schema in write process, use full schema in read process, for unspecified columns, use default value.

### PR DESCRIPTION
**Pull request by Linkerist**
Since partial update syntax is supported in #1546, #1829, it needs to be implemented in the reading/write process.
Use partial schema in the writing process, use an entire schema in the reading process, use the default value for unspecified columns.

usage, for example, for this table:
```
CREATE TABLE test (
 `k1` date NOT NULL,
 `k2` datetime NOT NULL,
 `k3` string NOT NULL,
 `k4` string NOT NULL,
 `k5` boolean NOT NULL,
 `v1` tinyint NOT NULL,
 `v2` smallint NOT NULL default '0'
) ENGINE=OLAP
PRIMARY KEY(`k1`, `k2`, `k3`, `k4`, `k5`) 
COMMENT "OLAP" 
DISTRIBUTED BY HASH(`k1`, `k2`, `k3`, `k4`, `k5`) BUCKETS 3 
PROPERTIES ( "replication_num" = "1", "storage_format" = "v2" );
```
you can use
```
$ curl --location-trusted -u root: -T basic_types_data_without_null -XPUT -H columns:k1,k2,k3,k4,k5,v1 -H partial_update:true http://172.26.92.139:48030/api/test/test/_stream_load
```
After this PR.
resolve #2059

### CODE REVIEW

**Comment by Linkerist** [comment]
run starrocks_be_unittest

File: be/src/storage/rowset/rowset.h [hunk 1]
```diff
@@ -125,6 +125,7 @@ class Rowset : public std::enable_shared_from_this<Rowset> {
     Status load();
 
     const TabletSchema& schema() const { return *_schema; }
+    inline void set_full_tablet_schema(const TabletSchema* full_tablet_schema) { _schema = full_tablet_schema; }
```
Comment (decster) [inline be/src/storage/rowset/rowset.h:128]: just set_schema

File: be/src/storage/vectorized/delta_writer.cpp [hunk 2]
```diff
@@ -124,14 +124,49 @@ Status DeltaWriter::_init() {
     }
 
     RowsetWriterContext writer_context(kDataFormatV2, config::storage_format_version);
+
+    const std::vector<SlotDescriptor*> req_slot_descriptors_without_op_col(
+            (*_req.slots).begin(), (*_req.slots).end() - ((*_req.slots).back()->col_name() == "__op"));
```
Comment (decster) [inline be/src/storage/vectorized/delta_writer.cpp:129]: Could we avoid copying this vector?

File: be/src/storage/vectorized/delta_writer.cpp [hunk 3]
```diff
@@ -124,14 +124,49 @@ Status DeltaWriter::_init() {
...
+            const auto& slot_col_name = slot_descriptor->col_name();
+            std::size_t index = _tablet->field_index(slot_col_name);
+            if (index < 0) {
+                std::stringstream ss;
+                ss << "invalid column name: " << slot_col_name;
+                LOG(WARNING) << ss.str();
+                return Status::InternalError(ss.str());
+            }
+            _tablet->tablet_schema().column(index).to_schema_pb(partial_tablet_schema_pb.add_column());
+        }
```
Comment (decster) [inline be/src/storage/vectorized/delta_writer.cpp:155]: maybe create a new method for TabletSchema to create a new TabletSchema by specifying columns, and call that method instead.
std::shared_ptr<TabletSchema> TabletSchema::create(src_schema, column_names_or_ids)

File: be/src/storage/vectorized/delta_writer.cpp [hunk 4]
```diff
@@ -264,11 +299,14 @@ Status DeltaWriter::close_wait(google::protobuf::RepeatedPtrField<PTabletInfo>*
         return Status::InternalError("Fail to flush memtable");
     }
 
+    LOG(WARNING) << "close_wait: ";
+
```
Comment (decster) [inline be/src/storage/vectorized/delta_writer.cpp:303]: remove this

**Review by decster — CHANGES_REQUESTED** [review]

File: be/src/storage/rowset/rowset.h [hunk 5]
```diff
@@ -125,7 +125,7 @@ class Rowset : public std::enable_shared_from_this<Rowset> {
     Status load();
 
     const TabletSchema& schema() const { return *_schema; }
-    inline void set_full_tablet_schema(const TabletSchema* full_tablet_schema) { _schema = full_tablet_schema; }
+    inline void set_schema(const TabletSchema* full_tablet_schema) { _schema = full_tablet_schema; }
```
Comment (decster) [inline be/src/storage/rowset/rowset.h:128]: Also parameter name

File: be/src/storage/vectorized/delta_writer.cpp [hunk 6]
```diff
@@ -125,36 +125,23 @@ Status DeltaWriter::_init() {
...
-        partial_tablet_schema_pb.set_keys_type(_tablet->tablet_schema().keys_type());
-        partial_tablet_schema_pb.set_bf_fpp(_tablet->tablet_schema().bf_fpp());
-        if (_tablet->tablet_schema().has_bf_fpp()) {
-            partial_tablet_schema_pb.set_bf_fpp(_tablet->tablet_schema().bf_fpp());
-        }
-        for (const auto* slot_descriptor : req_slot_descriptors_without_op_col) {
-            const auto& slot_col_name = slot_descriptor->col_name();
+    if (auto begin = (*_req.slots).begin(), end = (*_req.slots).end() - ((*_req.slots).back()->col_name() == "__op");
+        _tablet->tablet_schema().keys_type() == KeysType::PRIMARY_KEYS &&
+        (end - begin) < _tablet->tablet_schema().num_columns()) {
```
Comment (decster) [inline be/src/storage/vectorized/delta_writer.cpp:131]: Could you make it cleaner?

**Review by decster — DISMISSED** [review]

**Review by decster — DISMISSED** [review]

File: be/src/storage/vectorized/delta_writer.cpp [hunk 7]
```diff
@@ -124,14 +124,39 @@ Status DeltaWriter::_init() {
     }
 
     RowsetWriterContext writer_context(kDataFormatV2, config::storage_format_version);
+
+    std::size_t partial_col_num = (*_req.slots).size();
```
Comment (chaoyli) [inline be/src/storage/vectorized/delta_writer.cpp:128]: _req.slots->size() is ok

File: be/src/storage/vectorized/delta_writer.cpp [hunk 8]
```diff
@@ -124,14 +124,39 @@ Status DeltaWriter::_init() {
...
+    if (partial_col_num > 0 && (*_req.slots).back()->col_name() == "__op") {
+        --partial_col_num;
+    }
+    // maybe partial update, change to partial tablet schema
+    if (_tablet->tablet_schema().keys_type() == KeysType::PRIMARY_KEYS &&
+        partial_col_num < _tablet->tablet_schema().num_columns()) {
+        std::vector<std::size_t> column_indexes;
+        for (auto i = 0; i < partial_col_num; ++i) {
```
Comment (chaoyli) [inline be/src/storage/vectorized/delta_writer.cpp:136]: for (auto & slot_desc : *_req.slots)

**Comment by Linkerist** [comment]
run starrocks_be_unittest

**Review by decster — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**